### PR TITLE
fix: disable simplification for expression losses

### LINKED ISSUE #673: [BUG]: Simplification can leave expressions paired with stale cached losses
### What happened?

Below is a summary from Codex that is probably more lengthy than necessary. I think it might be useful for passing to a coding agent. 

The basic issue is pretty straightforward. We assume that after applying a simplification mutation the loss is unchanged but floating-point reassociation can change the actual loss, more so for some loss functions than others. This stale cached loss value then propagates around. I'm not sure that we want to change things to always recompute loss after a simplification since for many cases it will not make much of a difference but maybe we want some sort of option to turn that behavior on?

## Summary

SymbolicRegression assumes algebraic simplification does not change an
expression's loss. Simplified candidates can therefore retain their parent's
cached loss without being reevaluated.

This assumption is not always valid under floating-point arithmetic. In
particular, custom losses involving finite-difference derivatives can amplify
small numerical changes caused by reassociation and constant folding. The
resulting hall-of-fame entry can contain an expression whose recorded loss does
not match its actual loss.

## Observed behavior

This was found while using a custom `loss_function_expression` that measures
normalized tangency using central finite differences on `Float32` data.

At iteration 445:

- Expression complexity: 49
- Recorded loss: `0.039501722902059555`
- Recomputed loss: `0.03950172449891225`
- Difference: `1.6e-9`

By iteration 450, the expression had been algebraically simplified:

- Expression complexity: 47
- Recorded loss: `0.039501722902059555`
- Recomputed loss: `0.04476651308796335`
- Difference: `0.005264790185903798`

The cached loss remained bit-for-bit unchanged while the tree changed.

The unsimplified candidate reconstructs and reproduces its loss, which rules
out candidate serialization or printed constant precision as the cause. The
simplified expression contains reassociated additions and multiplications and
combined constants. These are algebraically equivalent transformations, but
they change Float32 evaluation order. The finite-difference calculation
amplifies the resulting rounding differences.

Across four 500-iteration searches, the observed scope was:

| Search | Mismatched Pareto candidates |
|---|---:|
| Real data, full grammar | 0/33 |
| Real data, no-trig grammar | 0/24 |
| Synthetic data, full grammar | 0/36 |
| Synthetic data, no-trig grammar | 2/23 |

Although uncommon in this example, stale losses can affect hall-of-fame
membership, candidate selection, and subsequent search behavior.

### Simplify mutation

In `src/Mutate.jl`, the `mutate!(..., ::Simplify, ...)` implementation:

1. Simplifies the tree.
2. Calculates the new complexity.
3. Constructs a child using `parent_member.loss`.
4. Returns immediately without calling `eval_cost`.

The implementation explicitly assumes simplification cannot change loss.

### Population-wide simplification

In `src/SingleIteration.jl`, `optimize_and_simplify_population` simplifies
population members before calling `finalize_costs`.

In `src/Population.jl`, `finalize_costs` currently recalculates losses only
when batching is enabled:

```julia
need_recalculate = options.batching
```

Consequently, population-wide simplification can also retain stale losses when
`batching=false`.

## Suggested fix

For the `Simplify` mutation, return the simplified tree through the normal
mutation evaluation path rather than constructing an immediate child with the
parent's loss. Conceptually:

```julia
simplify_tree!(new_tree, options.operators)
new_tree = combine_operators(new_tree, options.operators)
@recorder recorder["type"] = "simplify"
return MutationResult{N,P}(; tree=new_tree)
```

The normal path will then call `eval_cost` and construct the child using the
recomputed loss.

Population-wide simplification should also trigger loss recalculation. The
smallest correctness change would be:

```julia
need_recalculate = options.batching || options.should_simplify
```

If the additional evaluations are considered too expensive, a more selective
implementation could track whether simplification actually changed each tree
and reevaluate only those members.

## Suggested tests

1. Force a `Simplify` mutation on a Float32 expression with an
   evaluation-order-sensitive custom loss.
2. Verify that the returned member's loss equals a fresh `eval_loss` call.
3. Run population-wide simplification with `batching=false` and verify that
   every simplified member's cached loss matches its current tree.
4. Verify that losses exposed by `full_report` reproduce when its equations
   are reevaluated.

## Workaround

Setting `should_simplify=false` avoids the inconsistency, but it also disables
useful simplification and may produce unnecessarily large expressions.

### Version

2.0.0-alpha.11

### Operating System

Linux

### Interface

Julia REPL

### Relevant log output

```shell

```

## Minimal reproducer

The following standalone script forces a `Simplify` mutation on a small
Float32 expression and compares the cached loss with a fresh evaluation of the
simplified tree:

```julia
using SymbolicRegression
using SymbolicRegression: Dataset, RecordType
using SymbolicRegression.LossFunctionsModule: eval_loss
using SymbolicRegression.MutateModule: mutate!

function finite_difference_loss(tree, dataset, options)
    h = 1.0f-3
    Xp = copy(dataset.X)
    Xm = copy(dataset.X)
    Xp[1, :] .+= h
    Xm[1, :] .-= h

    yp, okp = eval_tree_array(tree, Xp, options)
    ym, okm = eval_tree_array(tree, Xm, options)
    (okp && okm) || return Float32(Inf)

    derivative = (only(yp) - only(ym)) / (2h)
    return abs(derivative)
end

options = Options(;
    binary_operators=(+, -, *),
    unary_operators=(),
    loss_function=finite_difference_loss,
    should_simplify=true,
)
@extend_operators options

dataset = Dataset(reshape(Float32[1000], 1, :), Float32[0])
x = Node{Float32}(; feature=1)
tree = (x * 100_000.0f0) * 1.0f-5
member = PopMember(dataset, tree, options; deterministic=true)

result = mutate!(
    copy(member.tree),
    member,
    Simplify(),
    options;
    recorder=RecordType(),
    dataset,
    parent_ref=1,
)
simplified = result.member
recomputed_loss = eval_loss(simplified.tree, dataset, options)

println("original tree:         ", member.tree)
println("simplified tree:       ", simplified.tree)
println("original loss:         ", member.loss)
println("cached simplified:     ", simplified.loss)
println("recomputed simplified: ", recomputed_loss)
println("absolute difference:   ", abs(recomputed_loss - simplified.loss))

@assert simplified.loss == member.loss
@assert abs(recomputed_loss - simplified.loss) > 0.03f0
```

Observed output:

```text
original tree:         binary_operator[3](binary_operator[3](x1, 100000.0), 1.0e-5)
simplified tree:       binary_operator[3](x1, 1.0)
original loss:         0.94604486
cached simplified:     0.94604486
recomputed simplified: 0.97656244
absolute difference:   0.030517578
```

### PR DESCRIPTION
**Pull request by MilesCranmerBot**
Fixes #673

Disable automatic simplification for `loss_function_expression`, matching `loss_function`.

### CODE REVIEW

File: src/Options.jl [hunk 1]
```diff
@@ -872,6 +872,7 @@ $(OPTION_DESCRIPTIONS)
     if should_simplify === nothing
         should_simplify = (
             loss_function === nothing &&
+            loss_function_expression === nothing &&
```
Comment (MilesCranmer) [inline src/Options.jl:875]: @adil-soubki ah it was this dumb human mistake of mine... See^.

**Review by MilesCranmer — APPROVED** [review]
